src/generic/C1_helper.cc

```diff
@@ -0,0 +1,379 @@
+    
+    // Add the constraining element to the mesh
+    constraint_mesh_pt->add_element_pt(constraint_element_pt);
+   }   
+ }
+
+
+
+//==============================================================================
+/// A function that upgrades straight sided elements to be curved. This involves
+/// Setting up the parametric boundary, F(s) and the first derivative F'(s)
+/// We also need to set the edge number of the upgraded element and the positions
+/// of the nodes j and k (defined below) and set which edge (k) is to be exterior
+///            @ k               
+///           /(                 
+///          /. \                
+///         /._._)               
+///      i @     @ j             
+/// For RESTING or FREE boundaries we need to have a C2 CONTINUOUS boundary
+/// representation. That is we need to have a continuous 2nd derivative defined
+/// too. This is well discussed in by [Zenisek 1981] (Aplikace matematiky ,
+/// Vol. 26 (1981), No. 2, 121--141). This results in the necessity for F''(s)
+/// as well.
+//=============================================================================
+ void upgrade_edge_elements_to_curved_boundaries(
+  Mesh* bulk_mesh_pt, 
+  std::map<unsigned,C1CurviLine*> c1_curviline_pt) 
+ {
+  
+  // Loop over the curvilinear parts of the outer boundary
+  for (const auto& [ibound, c1_curve_pt] : c1_curviline_pt)
+   {
+    
+    // Loop over the bulk elements adjacent to boundary ibound
+    const unsigned n_els=bulk_mesh_pt->nboundary_element(ibound);
+    for(unsigned e=0; e<n_els; e++)
+     {
+      // Get pointer to bulk element adjacent to b
+      FiniteElement* bulk_el_pt =
+       bulk_mesh_pt->boundary_element_pt(ibound,e);
+
+      // Initialise enum for the curved edge
+      C1Helper::CurvedEdgeEnumeration edge=C1Helper::CurvedEdgeEnumeration::none;
+      
+      // Loop over all (three) vertex nodes of the element and
+      // identify single node that is interior (i.e. not on any
+      // of the outer boundaries
+      unsigned index_of_interior_node = 3;
+      unsigned nnode_not_on_any_outer_boundary = 0;
+      const unsigned nnode = 3;
+      Vector<Vector<double> > xn(nnode,Vector<double>(2,0.0));
+      for(unsigned n=0;n<nnode;++n)
+       {
+        Node* nod_pt = bulk_el_pt->node_pt(n);
+        xn[n][0]=nod_pt->x(0);
+        xn[n][1]=nod_pt->x(1);
+        
+        // Check if it is on any of the outer boundaries
+        bool node_is_on_some_outer_boundary=false;
+        for (const auto& [b, dummy_c1_curve_pt] : c1_curviline_pt)
+         {
+          if (nod_pt->is_on_boundary(b))
+           {
+            node_is_on_some_outer_boundary=true;
+            break;
+           }
+         }
+        if (!node_is_on_some_outer_boundary)
+         {
+          index_of_interior_node = n;
+          nnode_not_on_any_outer_boundary++;
+         }
+       }// end record boundary nodes
+      
+      // hierher shouldn't these be called zeta (everywhere; sigh)
```

Comment:
Maybe. s_ubar and s_obar were chosen to be consistent with the literature. They are end coordinates, not the variable.

Reply:
rename.